consensus/istanbul/backend/handler.go

```diff
@@ -286,3 +301,154 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
+	} else {
+		go sendHandshake()
+	}
+
+	timeout := time.NewTimer(handshakeTimeout)
+	defer timeout.Stop()
+	select {
+	case err := <-errCh:
+		return false, err
+	case <-timeout.C:
+		return false, p2p.DiscReadTimeout
+	case isValidator := <-isValidatorCh:
+		return isValidator, nil
+	}
+}
+
+// readValidatorHandshakeMessage reads a validator handshake message.
+// Returns if the peer is a validator or if an error occurred.
+func (sb *Backend) readValidatorHandshakeMessage(peer consensus.Peer) (bool, error) {
+	logger := sb.logger.New("func", "readValidatorHandshakeMessage")
+	peerMsg, err := peer.ReadMsg()
+	if err != nil {
+		return false, err
+	}
+	if peerMsg.Code != istanbulValidatorHandshakeMsg {
+		logger.Warn("Read incorrect message code", "code", peerMsg.Code)
+		return false, errors.New("Incorrect message code")
+	}
+
+	var payload []byte
+	if err := peerMsg.Decode(&payload); err != nil {
+		return false, err
+	}
+
+	var msg istanbul.Message
+	err = msg.FromPayload(payload, sb.verifyValidatorHandshakeMessage)
+	if err != nil {
+		return false, err
+	}
+	// If the Signature is empty, the peer has decided not to reveal its info
+	if len(msg.Signature) == 0 {
+		return false, nil
+	}
+
+	var enodeCertificate enodeCertificate
+	err = rlp.DecodeBytes(msg.Msg, &enodeCertificate)
+	if err != nil {
+		return false, err
+	}
+
+	node, err := enode.ParseV4(enodeCertificate.EnodeURL)
+	if err != nil {
+		return false, err
+	}
+
+	// Ensure the node in the enodeCertificate matches the peer node
+	if node.ID() != peer.Node().ID() {
+		logger.Warn("Peer provided incorrect node ID in enodeCertificate", "enodeCertificate enode url", enodeCertificate.EnodeURL, "peer enode url", peer.Node().URLv4())
+		return false, errors.New("Incorrect node in enodeCertificate")
+	}
+
+	block := sb.currentBlock()
+	valSet := sb.getValidators(block.Number().Uint64(), block.Hash())
+	if !valSet.ContainsByAddress(sb.ValidatorAddress()) {
+		logger.Trace("This validator is not in the validator conn set")
+		return false, nil
+	}
+	if !valSet.ContainsByAddress(msg.Address) {
+		logger.Debug("Received a validator handshake message from peer not in the validator conn set", "msg.Address", msg.Address)
+		return false, nil
+	}
+
+	// If the enodeCertificate message is too old, we don't count the msg as valid
+	// An error is given if the entry doesn't exist, so we ignore the error.
+	knownVersion, err := sb.valEnodeTable.GetVersionFromAddress(msg.Address)
+	if err == nil && enodeCertificate.Version < knownVersion {
+		logger.Debug("Received a validator handshake message with an old version", "received version", enodeCertificate.Version, "known version", knownVersion)
+		return false, nil
+	}
+
+	// By this point, we know the peer is a validator and we update our val enode table accordingly
+	// Upsert will only use this entry if the version is new
+	err = sb.valEnodeTable.Upsert(map[common.Address]*vet.AddressEntry{msg.Address: {Node: node, Version: enodeCertificate.Version}})
```

Comment:
I thought a lot about this, and I think that if this node is a proxy, then it should not do the `upsert` here.  I think that the proxy should only send the certificate to the proxied validator, which will in turn update its valEnodeTable and then share that with its proxy.

This will  make reasoning of proxies' valEnodeTable updates/inserts much easier (e.g it's always the proxied validator that is driving the updates).  Also, when there are multiple proxies for a validator, doing it this way will prevent any race conditions on proxies valenodetables, since there would be only the proxied validator that is updating it and not both the proxied validator and the remote validator.

If we decide to make this change, this also implies that even though the remote validator sent the certificate to the proxy, the proxy should not return `true`, since there is a small possibility that the wrong proxy received it.

Reply:
Good point! I'm definitely in favor of this to keep the proxied validator in control. I think the issue of a validator connecting to the wrong proxy in the multiproxy setup is definitely a real issue. I'm also in support of changing the val enode share msg to be sent whenever the val enode table is updated to avoid the case that a proxy doesn't receive the entry until the timer ticks (up to a minute)

Reply:
I thought about this some more, and I think sending a val enode share msg when upserting is probably a bad idea. These messages aren't small by any means and it could flood proxied validator <-> proxy communication if someone decided to repeatedly increase their version and send an enodeCertificate. What do you think about the proxy sending the enodeCertificate to the proxied validator, and then having the proxied validator send it back to the proxy if it wants to? That way the race condition is solved and we don't risk flooding the proxy with a ton of data

Reply:
For some reason I thought the val enode share msg would cause the proxy to replace its val enode table with the proxied validator's table, but looking at the code it seems to be an `Upsert` that's done, so that eliminates the downsides of a race where:

1. Proxy upserts new entry from an enodeCertificate
2. Proxied validator sends val enode share msg
3. Proxied validator receives forwarded enodeCertificate
4. Proxy now receives val enode share msg

Trying to avoid letting multiproxy implementation details derail this PR too much, but IMO I think in the multiproxy setup the proxied validator should be in charge of which proxy will be used to connect to a new remote validator (and therefore should dictate the val enode table of the proxies).

I think in the single proxy setup it doesn't ultimately matter too much if the proxy upserts or not, but in the multiproxy setup we'll have to give this a deeper think. Are you ok with for the time being having the proxy give the proxied validator the authority to upsert, and then we can think this through more when we come out with multiproxy?

Another benefit to giving the authority to the proxied validator is if a proxy has been removed and a remote validator tries to connect to it. If the decision is left up to the proxied validator, then the remote validator will not incorrectly connect to a removed proxy.